shared/js/l10n.js

```diff
@@ -1607,27 +1684,32 @@
       }
     }
 
-    if (!this.ctx.availableLocales.length) {
-      // if there was no availableLanguages meta,
-      // register the default locale only
-      this.ctx.registerLocales(this.ctx.defaultLocale);
+    if (navigator.mozApps && navigator.mozApps.getAdditionalLanguages) {
+      navigator.mozApps.getAdditionalLanguages().then(
+        registerLocales.bind(this, meta));
+    } else {
+      registerLocales.call(this, meta);
     }
-    return initLocale.call(this);
```

Comment:
Ah, you're right, I didn't consider the fact that `getAdditionalLanguages` was async.  Now that you move most of the `available`/`extra` logic into the helper `registerLocales`, it makes a lot of sense to keep it around :)  Let's not make `initResources` any bigger.

Would it make sense to split this into smaller pieces maybe?  Like so:

```
if (navigator.mozApps && navigator.mozApps.getAdditionalLanguages) {
  navigator.mozApps.getAdditionalLanguages().then(
    extendLocales.bind(null, meta)).then(
      Function.prototype.apply.bind(this.ctx.registerLocales, this.ctx)).then(
        initLocale.bind(this));
} else {
  this.ctx.registerLocales(meta.defaultLocales, Object.keys(meta.availableLocales));
  initLocale.call(this);
}
```

…where `extendLocales` is your `registerLocales` function from line 1674 without its two last lines, returning `[defaultLanguage, availableLanguages]`.